src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+                } else {
+                    // our copy is still fresh
+                    resolved.insert(naddr.clone(), neighbor);
+                }
+                continue;
+            }
+
+            // need to resolve this one, but don't talk to it if we did so recently (even
+            // if we have stale information for it -- the remote node could be trying to trick
+            // us into DDoS'ing this node).
+            let peer_opt = PeerDB::get_peer(dbconn, network_id, &naddr.addrbytes, naddr.port)
+                .map_err(net_error::DBError)?;
+
+            if let Some(n) = peer_opt {
+                // we know about this neighbor, but its key didn't match the
+                // neighboraddress.  Only try to re-connect with it if we haven't done
+                // so recently, so a rogue neighbor can't force us to DDoS another
+                // peer.
+                if n.last_contact_time + NEIGHBOR_MINIMUM_CONTACT_INTERVAL < get_epoch_time_secs() {
+                    to_resolve.push((*naddr).clone());
+                } else {
+                    // recently contacted
+                    resolved.insert(naddr.clone(), n);
+                }
+            } else {
+                // okay, we really don't know about this neighbor
+                to_resolve.push((*naddr).clone());
+            }
+        }
+        Ok((resolved, to_resolve))
+    }
+
+    /// Select neighbors that are routable, and ignore ones that are not.
+    /// TODO: expand if we ever want to filter by unroutable network class or something
+    fn filter_sensible_neighbors(neighbors: Vec<NeighborAddress>) -> Vec<NeighborAddress> {
+        let mut ret = vec![];
+        for neighbor in neighbors.into_iter() {
+            if neighbor.addrbytes.is_anynet() {
+                continue;
+            }
+            ret.push(neighbor);
+        }
+        ret
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+
+        for slot in slots {
+            let peer_opt =
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
+                Some(_) => {
+                    return Ok(Some(slot));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Add a neighbor or schedule it to be pinged since it's up for replacement.
```

Comment:
Yes, that's correct.  If we have a new neighbor to add, but all of its slots are taken, then we can take one of two actions:

* Replace the old neighbor with the new neighbor
* Keep the old neighbor and drop the new neighbor

The code will prefer keeping old neighbors if they're still alive, since it prevents an evil neighbor eclipsing it by flooding them with garbage neighbor state.  To do this, the state machine will ping the old neighbor to see if it responds.  If it does, the new neighbor is dropped.  If it doesn't, the old neighbor is replaced.  This means that in order for an evil node to eclipse this node, it would have to set up lots and lots of real nodes for this node to handshake with as well as somehow sever all links to the old nodes.